# Bilimsel programlama

Bilimsel amaçla programlama yaparken, veri yalnızca bilgisayar ortamında işlenesi bir değer değildir. Artık anlamı olan bir değerdir.

Örneğin $25 {}^o C$'i bir sayı ile ifade edebiliriz. Fakat bu değerin artık bir anlamı var.

Dolayısyıla birimlere çok dikkat etmemiz gerek.

Bunu yapmak için çok dikkatlı olabiliriz. Kullanıcıyı verdiği değerler konusunda uyarabiliriz. Fakata ```Python```'da bu işin daha şık bir yolu var.

```astropy``` birim ve sabitler.

:::{warning}
```astropy```'ın sisteminizde kurulmuş olduğunu varsayıyorum.

Değilse:

```
$ pip install astropy
```
:::

## ```astropy.units``` (birimler)

```astropy.units``` elinizdeki verinin değeriyle işlenmesini ve gerektiğinde dönüşümler sağlar.

İçinde ne var ne yok bir bakalım.

In [40]:
from astropy import units

print(dir(units))

['A', 'AA', 'AB', 'ABflux', 'ABmag', 'AU', 'Angstrom', 'B', 'Ba', 'Barye', 'Bi', 'Biot', 'Bol', 'Bq', 'C', 'Celsius', 'Ci', 'CompositeUnit', 'D', 'DN', 'Da', 'Dalton', 'Debye', 'Decibel', 'DecibelUnit', 'Dex', 'DexUnit', 'EA', 'EAU', 'EB', 'EBa', 'EC', 'ED', 'EF', 'EG', 'EGal', 'EH', 'EHz', 'EJ', 'EJy', 'EK', 'EL', 'EN', 'EOhm', 'EP', 'EPa', 'ER', 'ERy', 'ES', 'ESt', 'ET', 'EV', 'EW', 'EWb', 'Ea', 'Eadu', 'Earcmin', 'Earcsec', 'Eau', 'Eb', 'Ebarn', 'Ebeam', 'Ebin', 'Ebit', 'Ebyte', 'Ecd', 'Echan', 'Ecount', 'Ect', 'Ed', 'Edeg', 'Edyn', 'EeV', 'Eerg', 'Eg', 'Eh', 'EiB', 'Eib', 'Eibit', 'Eibyte', 'Ek', 'El', 'Elm', 'Elx', 'Elyr', 'Em', 'Emag', 'Emin', 'Emol', 'Eohm', 'Epc', 'Eph', 'Ephoton', 'Epix', 'Epixel', 'Equivalency', 'Erad', 'Es', 'Esr', 'Eu', 'Evox', 'Evoxel', 'Eyr', 'F', 'Farad', 'Fr', 'Franklin', 'FunctionQuantity', 'FunctionUnitBase', 'G', 'GA', 'GAU', 'GB', 'GBa', 'GC', 'GD', 'GF', 'GG', 'GGal', 'GH', 'GHz', 'GJ', 'GJy', 'GK', 'GL', 'GN', 'GOhm', 'GP', 'GPa', 'GR', 'GRy', 'GS

```{image} ../images/yokamk.gif
:alt: Anaconda terminal
:class: bg-primary mb-1
:width: 400px
:align: center
```

Çokmuş. Ama arada tanıdıklar var. $metre$, $gram$, $second$ gibi. Bunlar üzerinden gidelim.

<hr>

Örnek

$3$ metrelik bir değer oluşturunuz.

In [15]:
from astropy import units

deger = 3 * units.meter
deger

<Quantity 3. m>

oluşan değerin "değeri" ve "birimi"ni gösteriniz

In [14]:
from astropy import units

deger = 3 * units.meter
print(deger.value)
print(deger.unit)

3.0
m


<hr>

Örnek

Yarıçapı $3$ $metre$ olan dairenin alanını hesaplayınız.

In [87]:
from astropy import units
from math import pi

r = 3 * units.meter
alan = pi * r ** 2
alan

<Quantity 28.27433388 m2>

<hr>

Örnek

Kenarları sırasıyla $75 cm$ ve $2 m$ olan dikdörtgenin alanını hesaplayınız.

In [39]:
from astropy import units


a = 75 * units.centimeter
b = 2 * units.meter

alan = a * b
alan

<Quantity 150. cm m>

$150 ~ cm ~ m$ ne biçin değerdir ya hu? Aslında değerde bir sorun yok. Sadece alışık değiliz böyle gösterimlere. Alışık olduğumuz hale getirmek için bu değeri $m^2$'ye veya temel birimlere dönüştürmek gerek.

### Dönüşüm

Her astropy ```units``` nesnesinin bir ```to``` metodu var. Bu metod bir birim nesnesi alır ve dönüşüm yapar.

<hr>

Örnek

Kenarları sırasıyla $75 cm$ ve $2 m$ olan dikdörtgenin alanını $m^2$ cinsinden hesaplayınız.

In [38]:
from astropy import units


a = 75 * units.centimeter
b = 2 * units.meter

alan = a * b
alan.to(units.m**2)

<Quantity 1.5 m2>

### Temel birimlere dönüşüm

Bir birim veya birimlerin kombinasyonları çok farklılık gösterebilir. Lakin ne kadar karmaşık olursa olsun bir sonucu temel birimler cinsinden görmek isteyebiliriz. ```astropy.units```'da temel birim, $mks$'dir.

```astropy.units``` temel birimlere dönüşüm için ```decompose``` metodunu kullanır.

<hr>

Örnek

$10 mil$'lik mesafeyi $25$ dakikada kateden bir aracın hızını $mks$ cinsinden hesaplayınız. 

In [37]:
from astropy import units


yol = 10 * units.imperial.mile
sure = 25 * units.minute

hiz = yol / sure

hiz.decompose()

<Quantity 10.72896 m / s>

### Özel birimler

```astropy.units``` kullanarak yeni birimler tanımlayabilirsiniz. 

<hr>

Örnek

Birim zamanda, metre başına kilogram olarak bir birim tanımlayıp bu birime sahip bir değer oluşturunuz.

In [36]:
from astropy import units

birim = units.kilogram / (units.second * units.meter)

deger = 4 * birim
deger

<Quantity 4. kg / (m s)>

## ```astropy.constants``` (sabitler)

```astropy.constants``` astronomide işe yarayan bir takım sabitlerdir.

İçinde ne var ne yok bir bakalım.

In [43]:
from astropy import constants

print(dir(constants))

['Constant', 'EMConstant', 'G', 'GM_earth', 'GM_jup', 'GM_sun', 'L_bol0', 'L_sun', 'M_earth', 'M_jup', 'M_sun', 'N_A', 'R', 'R_earth', 'R_jup', 'R_sun', 'Ryd', '__builtins__', '__cached__', '__doc__', '__file__', '__loader__', '__name__', '__package__', '__path__', '__spec__', 'a0', 'alpha', 'atm', 'au', 'b_wien', 'c', 'cgs', 'codata', 'codata2018', 'config', 'constant', 'e', 'eps0', 'g0', 'h', 'hbar', 'iau2015', 'iaudata', 'k_B', 'kpc', 'm_e', 'm_n', 'm_p', 'mu0', 'muB', 'pc', 'set_enabled_constants', 'si', 'sigma_T', 'sigma_sb', 'u', 'utils']


Sabitlerin sayısı daha azmış...

Bir tanesine bakalım.

<hr>

Örnek

Işık hızının değerini, birimini ve tanımını gösteriniz.

In [52]:
from astropy import constants


print(constants.c.value)
print(constants.c.unit)
print(constants.c)

299792458.0
m / s
  Name   = Speed of light in vacuum
  Value  = 299792458.0
  Uncertainty  = 0.0
  Unit  = m / s
  Reference = CODATA 2018


<hr>

Örnek

$E = m c^2$'yi kullanarak, $1 gr$ maddenin ne kadar enerji üretebildiğini, $erg$ cinsinden, gösteriniz.

In [88]:
from astropy import constants
from astropy import units

kutle = 1 * units.gram
enerji = kutle * constants.c ** 2
enerji.to(units.erg)

<Quantity 8.98755179e+20 erg>

<hr>

Örnek

Wien kayma yasasını kullanarak, $25{}^o K$ sıcaklığındaki bir cismin maksimum dalgaboyunu, Angstrom cinsinden, hesaplayan fonksiyonu yazınız.

Wien Kayma yasası

$$
\lambda_{max} = \frac{b}{T}
$$

:::{note}
$\lambda_{max}$ maksimum dalgaboyu, $b$ Wien kayma sabiti, $T$ ise kelvin cinsinden sıcaklığı ifade eder.
:::

şeklinde hsaplanır.

In [82]:
from astropy import constants
from astropy import units


def wien(T):
    return constants.b_wien / T

print(wien(25 * units.K).to(units.angstrom))

1159108.782074069 Angstrom


:::{warning}
Yazdığımız fonksiyon, sıcaklık değerini Kelvin olarak bekler. Celsius veya Fahrenheit, cinsinden sıcaklık değeri verilirse, bu fonksiyon hata verir. Dolayısıyla, verilen sıcaklık değerinin türüne bakmaksızın, değeri Kelvin'e dönüştürmekte yarar var.
:::

Sıcaklık dönüşümleri gibi, dönüşümlerde ```Equivalencies``` kullanma zorunluluğu vardır. Burada yine ```to``` metodu kullanılır ve dönüştütülmesi sitenen birim sağlanır. Fakat ek olarak ```equivalencies``` yani dönüşümü yapabilen nesne verilmelidir.

<hr>

Örnek

$25$ Fahrenheit değerini Kelvine'e dönüştürünüz.

In [83]:
from astropy import units

t = 25 * units.imperial.deg_F
t.to(units.K)

UnitConversionError: 'deg_F' (temperature) and 'K' (temperature) are not convertible

Gördüğünüz gibi $deg\_F$'den $K$'ye dönüşüm yapamadından yakındı.

In [84]:
from astropy import units

t = 25 * units.imperial.deg_F
t.to(units.K, equivalencies=units.temperature())

<Quantity 269.26111111 K>

Şimdi fonksiyonumuzu düzenleyelim:

In [86]:
from astropy import constants
from astropy import units


def wien(T):
    t_k = T.to(units.K, equivalencies=units.temperature())
    return constants.b_wien / t_k

print(wien(25 * units.K).to(units.angstrom))
print(wien(25 * units.deg_C).to(units.angstrom))
print(wien(25 * units.imperial.deg_F).to(units.angstrom))

1159108.782074069 Angstrom
97191.74761647401 Angstrom
107619.40122832671 Angstrom


```{image} ../images/cfk.png
:alt: Anaconda terminal
:class: bg-primary mb-1
:width: 400px
:align: center
```